ucan-discussion.md

```diff
@@ -0,0 +1,146 @@
+3. Permanent
+
+The one disadvantage is:
+* It results in having service specific data in the UCAN, which means it’s not unique across both web3.storage and nft.storage and it means it’s not reusable across different services	
+
+
+### Option 2 - User’s public key
+
+Another option is to use the user’s public key (or the first public key which they use with the service).
+
+This works well because it’s:
+1. Unique globally (across both services)
+
+Although it’s only unique across both services if the user uses the same key-pair for both services.
+
+A minor downside of tying the “bucket” ID to the user’s key-pair is that we’re creating a possibly confusing situation where a cryptographic identifier is being used for something which is really just a file path. Conceptually this might just be slightly odd to the user.
+
+There are two possible ways we could implement this:
+
+#### Option 2.1 - not stored in DB
+
+Use the user’s public key as the bucket name, but don’t store it in the DB.
+
+This is pretty much un-viable because:
+* If we don’t store it in the DB then we can’t make it unique to the user. So each time a user created a UCAN they could do so with a different public key, giving them (yet another) new bucket.
+* We have no way of knowing which bucket belongs to which user; we only know which bucket belongs to which key. This creates a situation with several distinct downsides:
+    - a. If the user loses their private key, they lose access to their bucket; we can’t implement any form of recovery.
+    - b. If the user’s private key is leaked, then they have no way of revoking it, so access to their bucket is permanently breached.
+    - c. The user could potentially create “backup” keys by creating UCANs with unlimited expiry times which delegate full privileges to another key. This helps mitigate problem (a), but makes problem (b) more likely.
+    - d. With no possibility of revocation or recovery, it conceptually breaks the idea that, like passwords, key-pairs can be rotated.
+
+#### Option 2.2 - stored in DB
+
+In this implementation (which is what’s currently implemented), when a user uploads the root DID the first time, we store the relationship between the user adding the root UCAN and the the DID.
+
+This solves all the problems of option 2.1, or at least leaves open the possibility of solving them by adding revocation/recovery functionality. (But still shares the overall downside of option 2 mentioned at the top.)
+
+
+### Option 3 - A UUID
+
+Another option is to generate a UUID for each user to use as their bucket name.
+
+The advantages of this are that it’s:
+1. Short
+2. Globally unique across services
+3. Removes the link between key-pairs and the bucket name, allowing key-pairs to be rotated.
+
+This UUID could be synced between web3.storage and nft.storage to allow each user to have a single bucket identifier which is global. But unlike with option 2, the user doesn’t sync this value for us.
```

Comment:
I'm not convinced `UUID` is better than just using DID, because it can't be derived. That said, it is aligned with my overall sentiment that bucket id and users first did just happen to be the same there is nothing more to it.

I would still prefer something derived from first user DID e.g. it's CID or a multihash which can look different enough yet is derived as opposed to made up.

There is also some benefit, that I can't fully put my finger on in regards to mapping buckets to DIDs as they map to IPNS etc... But then again that all is possible if id is derived somehow as opposed to been just made up.

Reply:
I was going to encourage preferring a something like a did:key-encoded public key to use as the identifier. I think the only downside over UUIDs might be compute cost?